atomix/storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java

```diff
@@ -602,6 +602,9 @@ void resetTail(final long index) {
     for (final SegmentedJournalReader reader : readers) {
       if (reader.getNextIndex() >= index) {
         reader.reset(index);
+      } else {
+        // This is not thread safe https://github.com/zeebe-io/zeebe/issues/4198
+        reader.reset(reader.getNextIndex());
```

Comment:
this could be really problematic

Reply:
When we call this? Truncate? this is called only be followers right? then it might be more okaisch

Reply:
Yes. Truncate is only called by followers. In another PR, we already moved ZeebeRaftStateMachine to use raft context because of this thread-safe issue.